### ISSUE TITLE
Default public file system path set to ./files results in broken images

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2683]
### Steps to reproduce

Content -> Add -> Post -> enter title -> upload image -> Save

### Actual behavior

No image is shown when rendering the content. Looking at the page source, there is a dot in the path of the image. For example `http://my.site.net/./files/styles/large/public/field/image/my_image.gif`

### Expected behavior

The image should be rendered properly.

---
PR by @jlfranklin : https://github.com/backdrop/backdrop/pull/1892
PR by @quicksketch: https://github.com/backdrop/backdrop/pull/1893

**Comment by olafgrabienski** [comment]
@klonos Can't reproduce it on a Backdrop 1.6.3 site, using the image field.

Image path: `http://example.com/files/styles/large/public/field/image/dummy-image.png`

On which Backdrop version do you see the issue?

**Comment by quicksketch** [comment]
This is a problem with 1.x and 1.6.x head branches. I believe this stemmed from #1853. Backdrop can't handle a files directory specified as `./files`, it has to be just `files`. This problem appears to only be that the installer sets the wrong files path. You can "fix" the problem by vising the File System settings page and updating the path. But obviously we need to fix the installer itself.

**Comment by jlfranklin** [comment]
Hmm.. I'll add some tests to try to catch it.  I'm torn as to the right solution.  Do we validate the form to prevent `./files` (assuming it doesn't already), or do we clean up the URL in `l()`, `url()` or the image field?

**Comment by quicksketch** [comment]
I think the primary issue is just that we manually set the file path in `system_install()`:

```
  // Set the public files path to the directory with the settings.php file.
  config_set('system.core', 'file_public_path', conf_path() . '/files');
```

So while adding validation on the form would also help (right now it accepts `./files` without any problem), the simplest fix is just making this config_set() a little smarter.

**Comment by jlfranklin** [comment]
OK.  I'll put together a PR updating that `config_set()` late tonight or sometime over the weekend.

**Comment by jlfranklin** [comment]
I don't know if I _can_ add tests for this.  The PR updates the install code, which (near as I can tell) isn't covered by Simpletest, because an installed system is a pre-requisite for the tests to run.

**Comment by quicksketch** [comment]
You're right @jlfranklin, we can't add tests for this because the installer itself is not covered. SimpleTests include a special installation method, which is why our tests didn't fail.

The PR looks like it would work. We can simplify the implementation though, a regex is probably overkill when all we really need to do is check if `conf_path()` returns the special situation of just a dot `.`

Is there a way to reproduce the problem with tablesort? The PR includes some kind of fix for tablesort during the installer, but that doesn't seem related to this problem with file handling.

**Comment by quicksketch** [comment]
I filed a slightly simpler implementation that doesn't include the tablesort fix at https://github.com/backdrop/backdrop/issues/1893

@jlfranklin Is there some other use-case besides just the single-site default directory that we need to account for?

**Comment by jlfranklin** [comment]
Multi-site paths also get the `./` prefix, `./sites/example.org/files`.

**Comment by quicksketch** [comment]
Oh, they sure do.

How about a simple `str_replace('./', '', conf_path())` then?

**Comment by quicksketch** [comment]
Updated the PR at https://github.com/backdrop/backdrop/pull/1893 with the `str_replace()` approach.

**Comment by quicksketch** [comment]
I'm tagging this for 1.7.0 just so that we don't release without this fix in place. It's critical that we fix this before release (though this will be merged to 1.6.x as well)

**Comment by jlfranklin** [comment]
> How about a simple `str_replace('./', '', conf_path())` then?

That'll catch any `./` in the middle of a sites path.  Unlikely, to be sure, but what a head-scratcher that would be to debug.

Go ahead and put in the `str_replace()`.  I'm not sure what the objection is to the regex solution, but we can discuss it for 1.7.1, and this will fix the immediate issue.

**Comment by jlfranklin** [comment]
> Is there a way to reproduce the problem with tablesort? The PR includes some kind of fix for tablesort during the installer, but that doesn't seem related to this problem with file handling.

Try to install the system by hand with the files directory unwritable, or any other error that would trigger the status page.  The install stops just before the status page with an error about `tablesort_cell()` not defined.

**Comment by jlfranklin** [comment]
I ran across the tablesort issue when testing my PR.

**Comment by jlfranklin** [comment]
Should we add an update hook to look for `./` in the public files path?

**Comment by quicksketch** [comment]
> Should we add an update hook to look for ./ in the public files path?

I don't think it's necessary as we never had an official release with this bug present (other than the 1.7-preview release).

**Comment by jlfranklin** [comment]
#2693 is a report for the tablesort bug.

**Comment by quicksketch** [comment]
Merged backdrop/backdrop#1893 into 1.x and 1.6.x. The tablesort bug we'll fix in #2693.
